renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -0,0 +1,604 @@
+    regions.push_back(region);
+    aspectFlags = VkImageAspectFlags(VK_IMAGE_ASPECT_COLOR_BIT);
+  }
+  else
+  {
+    region.imageSubresource = VkImageSubresourceLayers{VK_IMAGE_ASPECT_DEPTH_BIT, 0, 0, 1};
+    if(IsDepthOnlyFormat(p.srcImageFormat) || IsDepthAndStencilFormat(p.srcImageFormat))
+    {
+      regions.push_back(region);
+      aspectFlags |= VK_IMAGE_ASPECT_DEPTH_BIT;
+    }
+    if(IsStencilFormat(p.srcImageFormat))
+    {
+      region.imageSubresource.aspectMask = VK_IMAGE_ASPECT_STENCIL_BIT;
+      region.bufferOffset = offset + 4;
+      regions.push_back(region);
+      aspectFlags |= VK_IMAGE_ASPECT_STENCIL_BIT;
+    }
+  }
+
+  VkImageMemoryBarrier barrier = {VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER,
+                                  NULL,
+                                  VK_ACCESS_SHADER_WRITE_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT |
+                                      VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT |
+                                      VK_ACCESS_TRANSFER_WRITE_BIT | VK_ACCESS_MEMORY_WRITE_BIT,
+                                  VK_ACCESS_TRANSFER_READ_BIT,
+                                  p.srcImageLayout,
+                                  VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL,
+                                  VK_QUEUE_FAMILY_IGNORED,
+                                  VK_QUEUE_FAMILY_IGNORED,
+                                  Unwrap(p.srcImage),
+                                  {aspectFlags, 0, 1, 0, 1}};
+
+  DoPipelineBarrier(cmd, 1, &barrier);
+
+  ObjDisp(cmd)->CmdCopyImageToBuffer(Unwrap(cmd), Unwrap(p.srcImage),
+                                     VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, Unwrap(p.dstBuffer),
+                                     (uint32_t)regions.size(), regions.data());
+
+  barrier.image = Unwrap(p.srcImage);
+  barrier.srcAccessMask = VK_ACCESS_TRANSFER_READ_BIT;
+  barrier.dstAccessMask = VK_ACCESS_TRANSFER_READ_BIT;
+  barrier.oldLayout = VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL;
+  barrier.newLayout = p.srcImageLayout;
+  DoPipelineBarrier(cmd, 1, &barrier);
+}
+
+void CreateOcclusionPool(VkDevice dev, uint32_t poolSize, VkQueryPool *pQueryPool)
+{
+  VkQueryPoolCreateInfo occlusionPoolCreateInfo = {VK_STRUCTURE_TYPE_QUERY_POOL_CREATE_INFO};
+  occlusionPoolCreateInfo.queryType = VK_QUERY_TYPE_OCCLUSION;
+  occlusionPoolCreateInfo.queryCount = poolSize;
+  // TODO: check that occlusion feature is available
+  VkResult vkr =
+      ObjDisp(dev)->CreateQueryPool(Unwrap(dev), &occlusionPoolCreateInfo, NULL, pQueryPool);
+  RDCASSERTEQUAL(vkr, VK_SUCCESS);
+}
+
+VkImageLayout VulkanDebugManager::GetImageLayout(ResourceId image, VkImageAspectFlags aspect,
+                                                 uint32_t mip)
+{
+  VkImageLayout imgLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+  const ImageLayouts &imgLayouts = m_pDriver->m_ImageLayouts[image];
+  for(const auto &resState : imgLayouts.subresourceStates)
```

Comment:
Avoid using `auto` except where the type is verbosely ugly, like lambdas or iterators.

Reply:
Done